modules/ensemble/lib/action/bluetooth_action.dart

```diff
@@ -0,0 +1,189 @@
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:get_it/get_it.dart';
+
+class InitializeBluetoothAction extends EnsembleAction {
+  BluetoothManager bluetoothManager = GetIt.I<BluetoothManager>();
+
+  EnsembleAction? onError;
+  EnsembleAction? onDataStream;
+
+  InitializeBluetoothAction(
+      {super.initiator, super.inputs, this.onError, this.onDataStream});
+
+  factory InitializeBluetoothAction.from(
+          {Invokable? initiator, dynamic payload}) =>
+      InitializeBluetoothAction.fromYaml(
+          initiator: initiator, payload: Utils.getYamlMap(payload));
+
+  factory InitializeBluetoothAction.fromYaml(
+      {Invokable? initiator, Map? payload}) {
+    return InitializeBluetoothAction(
+      initiator: initiator,
+      inputs: Utils.getMap(payload?['inputs']),
+      onError: EnsembleAction.from(payload?['onError'], initiator: initiator),
+      onDataStream:
+          EnsembleAction.from(payload?['onDataStream'], initiator: initiator),
+    );
+  }
+
+  Future<dynamic> execute(
+      BuildContext context, ScopeManager scopeManager) async {
+    if (kIsWeb && onError != null) {
+      return ScreenController().executeAction(context, onError!,
+          event: EnsembleEvent(
+            initiator,
+            error: 'Bluetooth is not supported on the web',
+            data: {'status': 'error'},
+          ));
+    }
+
+    // TODO: ON page pop cancel stream
+    // bluetoothManager.init(
+    //   context: context,
+    //   initiator: initiator,
+    //   onDataStream: onDataStream,
+    // );
+    await bluetoothManager.startScan(onScanResult: (data) {
+      if (onDataStream != null) {
+        ScreenController().executeAction(context, onDataStream!,
+            event: EnsembleEvent(
+              initiator,
+              data: data,
+            ));
+      }
+      return null;
+    });
+  }
+}
+
+class ConnectBluetoothAction extends EnsembleAction {
+  BluetoothManager bluetoothManager = GetIt.I<BluetoothManager>();
+
+  EnsembleAction? onError;
+  EnsembleAction? onDataStream;
+
+  final String deviceId;
+
+  ConnectBluetoothAction(
+      {super.initiator,
+      super.inputs,
+      this.onError,
+      this.onDataStream,
+      required this.deviceId});
+
+  factory ConnectBluetoothAction.from(
+          {Invokable? initiator, dynamic payload}) =>
+      ConnectBluetoothAction.fromYaml(
+          initiator: initiator, payload: Utils.getYamlMap(payload));
+
+  factory ConnectBluetoothAction.fromYaml(
+      {Invokable? initiator, Map? payload}) {
+    return ConnectBluetoothAction(
+      initiator: initiator,
+      inputs: Utils.getMap(payload?['inputs']),
+      onError: EnsembleAction.from(payload?['onError'], initiator: initiator),
+      onDataStream:
+          EnsembleAction.from(payload?['onDataStream'], initiator: initiator),
+      deviceId: Utils.getString(payload?['deviceId'], fallback: 'fallback'),
```

Comment:
should throw error if deviceId is not specified instead.

Reply:
yes, updating the negative flow right now.